generator.go

```diff
@@ -84,31 +86,61 @@ type Generator interface {
 	NewV5(ns UUID, name string) UUID
 }
 
-// Default generator implementation.
-type rfc4122Generator struct {
+// Gen is a reference UUID generator based on the specifications laid out in
+// RFC-4122 and DCE 1.1: Authentication and Security Services. This type
+// satisfies the Generator interface as defined in this package.
+//
+// For consumers who are generating V1 UUIDs, but don't want to expose the MAC
+// address of the node generating the UUIDs, the NewGenHWAF() function has been
+// provided as a convenience. See the function's documentation for more info.
+//
+// The authors of this package do not feel that the majority of users will need
+// to obfuscate their MAC address, and so we recommend using NewGen() to create
+// a new generator.
+type Gen struct {
 	clockSequenceOnce sync.Once
 	hardwareAddrOnce  sync.Once
 	storageMutex      sync.Mutex
 
 	rand io.Reader
 
 	epochFunc     epochFunc
-	hwAddrFunc    hwAddrFunc
+	hwAddrFunc    HWAddrFunc
 	lastTime      uint64
 	clockSequence uint16
 	hardwareAddr  [6]byte
 }
 
-func newRFC4122Generator() Generator {
-	return &rfc4122Generator{
+// interface check -- build will fail if *Gen doesn't satisfy Generator
+var _ Generator = (*Gen)(nil)
+
+// NewGen returns a new instance of Gen with some default values set. Most
+// people should use this.
+func NewGen() *Gen {
+	return NewGenHWAF(defaultHWAddrFunc)
+}
+
+// NewGenHWAF builds a new UUID generator with the HWAddrFunc provided. Most
+// consumers should use NewGen() instead.
+//
+// This is used so that consumers can generate their own MAC addresses, for use
+// in the generated UUIDs, if there is some concern about exposing the physical
+// address of the machine generating the UUID.
+//
+// The Gen generator will only invoke the HWAddrFunc once, and cache that MAC
+// address for all the future UUIDs generated by it. If you'd like to switch the
+// MAC address being used, you'll need to create a new generator using this
+// function.
+func NewGenHWAF(hwaf HWAddrFunc) *Gen {
```

Comment:
I'll be the first to admit that I'm not sold on this function name. Couldn't think of anything better in the moment, though.

Reply:
I personally tend to name things with `With` in such cases: `NewGenWithHWAF`

Reply:
That sounds better to me.